src/cockpit/samples.py

```diff
@@ -121,21 +132,27 @@ def detect_cpu_sensors(self, hwmonid: int, name: str) -> None:
                 if name != 'cpu_thermal':
                     continue
 
-            self.sensors.append(sensor_path)
+            # need full path
+            self.sensors.append(f'{HWMON_PATH}/{hwmon_name}/{sensor}')
 
     def sample(self, samples: Samples) -> None:
         cpu_names = ['coretemp', 'cpu_thermal', 'k8temp', 'k10temp', 'atk0110']
 
         if not self.sensors:
-            # TODO: 2 ** 32?
-            for index in range(0, 2 ** 32):
-                try:
-                    with open(f'/sys/class/hwmon/hwmon{index}/name') as f:
-                        name = f.read().strip()
-                    if name in cpu_names:
-                        self.detect_cpu_sensors(index, name)
-                except FileNotFoundError:
-                    break
+            with Handle.open(HWMON_PATH, os.O_RDONLY | os.O_DIRECTORY) as top:
+                for hwmon_name in os.listdir(top):
+                    with Handle.open(hwmon_name, os.O_RDONLY | os.O_DIRECTORY, dir_fd=top) as sub:
+                        try:
+                            fd = os.open('name', os.O_RDONLY, dir_fd=sub)
+                            try:
+                                name = os.read(fd, 1024)
+                                name = name.decode().strip()
+                                if name in cpu_names:
+                                    self.detect_cpu_sensors(name, hwmon_name, sub)
+                            finally:
+                                os.close(fd)
+                        except FileNotFoundError:
+                            continue
```

Comment:
We're once or twice too many times indented here.  I feel like this (highlighted section) should be a separate function.